### PR TITLE
Make DAC and ProfToEEInterfaceImpl stop using BaseDomain

### PR DESCRIPTION
**Pull request by elinor-fung**
`BaseDomain` should no longer be needed now that we only have one `AppDomain` and the `SystemDomain` can be treated as separate. This makes the DAC and ProfToEEInterfaceImpl use `AppDomain` directly and check against `SystemDomain::System()` to determine if a pointer is the system domain.

cc @jkotas @AaronRobinsonMSFT

### CODE REVIEW

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp [hunk 1]
```diff
@@ -5506,11 +5489,20 @@ HRESULT ProfToEEInterfaceImpl::GetAppDomainInfo(AppDomainID appDomainId,
...
         *pProcessId = 0;
 
+    // <TODO>@todo:
+    // Right now, the AppDomainID is not a true AppDomain, since we use the old
+    // AppDomain/SystemDomain model in the profiling API. This means that the
+    // profiler exposes the SystemDomain to the outside world. It's not clear
+    // whether this is actually the right thing to do or not. - seantrow
+    //
+    // Postponed to V2.
+    // </TODO>
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/vm/proftoeeinterfaceimpl.cpp:5499]: I think this can be removed.

**Review by AaronRobinsonMSFT — APPROVED** [review]
Can you try a few SOS scenarios under WinDBG just to validate the commands still work as intended.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by mikem8361 — APPROVED** [review]

**Review by davmason — APPROVED** [review]
The profiler side LGTM

File: src/coreclr/vm/proftoeeinterfaceimpl.cpp [hunk 2]
```diff
@@ -3148,9 +3148,9 @@ HRESULT ProfToEEInterfaceImpl::GetAppDomainStaticAddress(ClassID classId,
         return E_INVALIDARG;
     }
 
-    // Some domains, like the system domain, aren't APP domains, and thus don't contain any
+    // The system domain isn't an APP domain and thus doesn't contain any
     // statics.  See if the profiler is trying to be naughty.
-    if (!((BaseDomain*) appDomainId)->IsAppDomain())
+    if (appDomainId == (AppDomainID)SystemDomain::System())
```
Comment (jkotas) [inline src/coreclr/vm/proftoeeinterfaceimpl.cpp:3153]: How can the profiler ever get access to the `SystemDomain::System()` pointer? This looks like a unreachable code that can be deleted.

Comment (davmason) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:3153]: I remember seeing some weirdness around the system domain in the profiler APIs. I think they do get a callback when the system domain is created, but it's been a while so I am checking now

Comment (jkotas) [reply src/coreclr/vm/proftoeeinterfaceimpl.cpp:3153]: You are right: https://github.com/dotnet/runtime/blob/2ed43b6a8105563f491fb521d761e6eff6ac4448/src/coreclr/vm/appdomain.cpp#L1533-L1545

This should stay then.

**Review by jkotas — APPROVED** [review]

**Pull request merged**